File: backend/routes/filters.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
import json
from models import db, SavedFilter
from utils.response import success_response
from utils.error_handler import APIError

filters_bp = Blueprint('filters', __name__)

@filters_bp.route('', methods=['GET'])
@jwt_required()
def list_filters():
    user_id = get_jwt_identity()
    page_name = request.args.get('page_name')
    
    query = SavedFilter.query.filter_by(user_id=user_id)
    if page_name:
        query = query.filter_by(page_name=page_name)
    
    filters = query.order_by(SavedFilter.is_default.desc(), SavedFilter.created_at.desc()).all()
    
    return jsonify(success_response([f.to_dict() for f in filters]))

@filters_bp.route('', methods=['POST'])
@jwt_required()
def save_filter():
    user_id = get_jwt_identity()
    data = request.get_json()
    page_name = data.get('page_name')
    filter_name = data.get('filter_name')
    filter_data = data.get('filter_data')
    is_default = data.get('is_default', False)
    
    if not page_name or not filter_name or filter_data is None:
        raise APIError('请填写完整信息', 400)
    
    if is_default:
        SavedFilter.query.filter_by(
            user_id=user_id,
            page_name=page_name,
            is_default=True
        ).update({'is_default': False})
    
    saved_filter = SavedFilter(
        user_id=user_id,
        page_name=page_name,
        filter_name=filter_name,
        filter_data=json.dumps(filter_data) if isinstance(filter_data, dict) else filter_data,
        is_default=is_default
    )
    
    db.session.add(saved_filter)
    db.session.commit()
    
    return jsonify(success_response(saved_filter.to_dict(), '已保存筛选条件'))

@filters_bp.route('/<int:filter_id>', methods=['PUT'])
@jwt_required()
def update_filter(filter_id):
    user_id = get_jwt_identity()
    saved_filter = SavedFilter.query.get(filter_id)
    
    if not saved_filter or saved_filter.user_id != user_id:
        raise APIError('筛选条件不存在', 404)
    
    data = request.get_json()
    
    if 'filter_name' in data:
        saved_filter.filter_name = data['filter_name']
    if 'filter_data' in data:
        saved_filter.filter_data = json.dumps(data['filter_data']) if isinstance(data['filter_data'], dict) else data['filter_data']
    if 'is_default' in data:
        if data['is_default']:
            SavedFilter.query.filter_by(
                user_id=user_id,
                page_name=saved_filter.page_name,
                is_default=True
            ).update({'is_default': False})
        saved_filter.is_default = data['is_default']
    
    db.session.commit()
    
    return jsonify(success_response(saved_filter.to_dict(), '已更新'))

@filters_bp.route('/<int:filter_id>', methods=['DELETE'])
@jwt_required()
def delete_filter(filter_id):
    user_id = get_jwt_identity()
    saved_filter = SavedFilter.query.get(filter_id)
    
    if not saved_filter or saved_filter.user_id != user_id:
        raise APIError('筛选条件不存在', 404)
    
    db.session.delete(saved_filter)
    db.session.commit()
    
    return jsonify(success_response(None, '已删除'))

File: backend/src/routes/filters.js
import { Router } from 'express'
import prisma from '../prisma.js'
import { authMiddleware } from '../middleware/auth.js'

const router = Router()

router.get('/:pageKey', authMiddleware, async (req, res) => {
  const { roleDefault } = req.query
  const userFilters = await prisma.savedFilter.findMany({
    where: {
      userId: req.user.id,
      pageKey: req.params.pageKey
    },
    orderBy: [{ isDefault: 'desc' }, { createdAt: 'desc' }]
  })

  let roleDefaultFilter = null
  if (roleDefault) {
    roleDefaultFilter = await prisma.savedFilter.findFirst({
      where: {
        pageKey: req.params.pageKey,
        roleDefault: roleDefault
      },
      orderBy: { createdAt: 'desc' }
    })
  }

  res.json({ code: 0, data: { filters: userFilters, roleDefaultFilter } })
})

router.post('/', authMiddleware, async (req, res) => {
  const { pageKey, name, filterData, isDefault, roleDefault } = req.body

  if (isDefault) {
    await prisma.savedFilter.updateMany({
      where: { userId: req.user.id, pageKey, isDefault: true },
      data: { isDefault: false }
    })
  }

  if (roleDefault) {
    await prisma.savedFilter.updateMany({
      where: { pageKey, roleDefault },
      data: { roleDefault: null }
    })
  }

  const filter = await prisma.savedFilter.create({
    data: {
      userId: req.user.id,
      pageKey,
      name,
      filterData,
      isDefault: isDefault || false,
      roleDefault: roleDefault || null
    }
  })
  res.json({ code: 0, data: filter })
})

router.put('/:id', authMiddleware, async (req, res) => {
  const { name, filterData, isDefault, roleDefault } = req.body
  const id = parseInt(req.params.id)

  const existing = await prisma.savedFilter.findUnique({ where: { id } })

  if (isDefault && !existing.isDefault) {
    await prisma.savedFilter.updateMany({
      where: { userId: req.user.id, pageKey: existing.pageKey, isDefault: true },
      data: { isDefault: false }
    })
  }

  if (roleDefault && roleDefault !== existing.roleDefault) {
    await prisma.savedFilter.updateMany({
      where: { pageKey: existing.pageKey, roleDefault },
      data: { roleDefault: null }
    })
  }

  const updateData = { name, filterData, isDefault }
  if (roleDefault !== undefined) {
    updateData.roleDefault = roleDefault || null
  }

  const filter = await prisma.savedFilter.update({
    where: { id },
    data: updateData
  })
  res.json({ code: 0, data: filter })
})

router.delete('/:id', authMiddleware, async (req, res) => {
  await prisma.savedFilter.delete({ where: { id: parseInt(req.params.id) } })
  res.json({ code: 0 })
})

export default router
